bpf/process/bpf_execve_event.c

```diff
@@ -169,11 +169,18 @@ read_exe(struct task_struct *task, struct heap_exe *exe)
 	struct file *file = BPF_CORE_READ(task, mm, exe_file);
 	struct path *path = __builtin_preserve_access_index(&file->f_path);
 
-	exe->len = BINARY_PATH_MAX_LEN;
-	exe->off = (char *)&exe->buf;
-	exe->off = __d_path_local(path, exe->off, (int *)&exe->len, (int *)&exe->error);
-	if (exe->len > 0)
-		exe->len = BINARY_PATH_MAX_LEN - exe->len;
+	// we need to walk the complete 4096 len dentry in order to have an accurate
+	// matching on the prefix operators, even if we only keep a subset of that
+	char *buffer;
+
+	buffer = d_path_local(path, (int *)&exe->len, (int *)&exe->error);
+	if (!buffer)
+		return 0;
+
+	// buffer used by d_path_local can contain up to MAX_BUF_LEN i.e. 4096 we
+	// only keep the first 255 chars for our needs
+	exe->len &= 0xFF;
```

Comment:
Masking to the lower 255 bits seems wrong. Surely if returned path is 257 in length, then masking it with 0xFF will make the length 1, instead of 255.

Reply:
hehe you caught the bug I caught during the rhel8 investigation indeed what I wanted here, nice catch thanks!
```diff
-       exe->len &= 0xFF;
+       bpf_printk("exe before mask: %d", exe->len);
+       if (exe->len >= 256) {
+               exe->len = 256;
+       }
```

Reply:
I think it should be exe->len = 255; (instead of 256) unless the comment is wrong?

Reply:
I bumped the comment, why would you take 255 instead of 256?

Reply:
It might not matter here, but often in BPF we set upper limits as one less than a power of two, so that we can do "&= MASK" type operations to constrain an index into the array (to satisfy the verifier as opposed to avoiding overflows). Maybe check that this buffer doesn't need this constraint?

Reply:
Indeed I didn't check not sure why but this is what makes the verifier unhappy. It's unhappy mostly because `exe->len` seems to be possibly negative. If I add a naive `if v < 0; v = 0` it gets optimized out.

Reply:
exe->len is a u32 so the compiler is absolutely sure it can't be <0 :D hence it gets rid of it. The verifier on the other hand isn't so sure (and rightly so).
Solution is something like:
```
if (exe->len > 255) {
    exe->len = 255;
}
asm volatile("%[len] &= 0xff;\n"
     : [len] "+r"(exe->len)
     :);
```
i.e. limit it to 255 to get the right logic, and then mask it to make the verifier happy, using "asm volatile" to force the compiler to insert it at this point.

Reply:
Yes this way it works I tried but I lost 1 char, I tried to make it work with 256 :D

Reply:
It seems that the inline asm is no longer needed, do you think we should continue use it like that?

Reply:
You only need the inline asm if the verifier complains. It varies between kernel versions so 🤷 .

Reply:
So apparently can:
- sacrifice 1 char to the verifier god and continue my day
- write a potentially broken bitwise operation on 32 bits and fight it

Reply:
I think sacrificing 1 byte (which we do everywhere) is less complex and therefore easier to maintain.